Documentation/design-docs/arm64-intrinsics.md

```diff
@@ -0,0 +1,217 @@
+The opposite is also true.
+
+Any platform which does not implement a namespace, must:
+
++ Return `false` for `IsSupported`
++ `throw` on any unimplemented function
+
+#### Partial implementation of intrinsic namespace
+
++ It is preferred that `IsSupported` represents the state of an entire namespace
++ It is certainly required at time of a release, `IsSupported` represents the state of an entire namespace
++ Allow for writing tests and implementing code, `IsSupported == false` may be used when namespace is partially implemented
+
+#### Addition of new intrinsics to existing namespace
+When an intrinsic namespace is introduced, a best guess will be made about the set of useful intrinsics.
+However, the intrinsics will be added as needed or requested for various optimization efforts.
+
++ The set of intrinsics instructions supported by a namespace should only be allowed
+if all platforms reporting `IsSupported` are supported (at least at time of release)
++ It may be preferable to add new namespaces in cases where this creates significant burden
+
+### Half precision floating point
+
+This document will refer to half precision floating point as `Half`.
+
++ Machine learning and Artificial intelligence often use `Half` type to simplify storage and improve processing time.
++ CoreCLR and `CIL` in general do not have general support for a `Half` type
++ There is an open request to expose `Half` intrinsics
++ ARM64 will introduce `System.Runtime.Intrinsics.Half` to support this request 
++ `Half` will be defined as:
+
+```
+    [StructLayout(LayoutKind.Sequential, Size = 2)]
+    public struct Half : struct {}
+``` 
+
+#### ARMv8 Half precision support
+
+ARMv8 baseline support for `Half` is limited.  The following operations are supported
+
++ Loads and Stores
++ Conversion to/from `Float`
++ Widening from `Vector128<Half>` to two `Vector128<Float>`
++ Narrowing from two `Vector128<Float>` to `Vector128<Float>`
+
+It is presumed that even this minimal support could be helpful
+
+Optional `Half` extensions add more complete support
+
+#### `Half` and ARM64 ABI
+
+The proposed `Half` implementation will treat `Half` as raw bits.  
+It is likely therefore that this will not conform to the ARM64 ABI specifically with respect to the HFA
+calling convention.  This will hinder use of `Half` in `PInvoke` calls.  Resolving this ABI issue will
+be a low priority.
+
+### Scalable Vector Extension Support
+
+`SVE`, the Scalable Vector Extension introduces its own complexity.  
+
+The extension 
++ Creates a new set of SVE registers 
++ Each register has a platform specific length
+  + Any multiple of 128 bits
+
+Therefore implementation will not be trivial.
+
++ Register allocator may need changes
++ Crossgen of SVE intrinsics must be delayed until runtime JIT
++ SIMD support will face similar issues
++ Open issue: Should we use `Vector<T>` or SVE<T> in user interface design?
```

Comment:
Thinking in design discussions was to use Vector128&lt;T>, Vector256&lt;T>, Vector384&lt;T>, Vector512&lt;T>, ... Vector2048&lt;T>.

Reply:
The problem with that approach is the user is required to write many variants of identical code

Maybe `Vector<128, T>` would work better

Reply:
Nah, `Vector<128, T>` won't work, generics do not support non type arguments.

> The problem with that approach is the user is required to write many variants of identical code

They could make the code generic on the vector type. Even if `Vector<128, T>` would work they'd still need to do the same thing.

Reply:
There is a language proposal to extend generics with this functionality see: https://github.com/dotnet/csharplang/issues/749